File: wrappers/go/rank_cache.go
package turbotoken

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sync"
)

var (
	cacheMu sync.Mutex
)

// CacheDir returns the directory used for caching rank files.
// Uses TURBOTOKEN_CACHE_DIR if set, otherwise ~/.cache/turbotoken/.
func CacheDir() string {
	if dir := os.Getenv("TURBOTOKEN_CACHE_DIR"); dir != "" {
		return dir
	}
	home, err := os.UserCacheDir()
	if err != nil {
		home = os.TempDir()
	}
	return filepath.Join(home, "turbotoken")
}

// EnsureRankFile ensures the rank file for the given encoding exists locally,
// downloading it if necessary. Returns the path to the local file.
func EnsureRankFile(name string) (string, error) {
	spec, err := GetEncodingSpec(name)
	if err != nil {
		return "", err
	}

	dir := CacheDir()
	localPath := filepath.Join(dir, name+".tiktoken")

	cacheMu.Lock()
	defer cacheMu.Unlock()

	if _, err := os.Stat(localPath); err == nil {
		return localPath, nil
	}

	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", fmt.Errorf("turbotoken: create cache dir: %w", err)
	}

	resp, err := http.Get(spec.RankFileURL)
	if err != nil {
		return "", downloadError(spec.RankFileURL, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", downloadError(spec.RankFileURL, fmt.Errorf("HTTP %d", resp.StatusCode))
	}

	tmp := localPath + ".tmp"
	f, err := os.Create(tmp)
	if err != nil {
		return "", fmt.Errorf("turbotoken: create temp file: %w", err)
	}

	if _, err := io.Copy(f, resp.Body); err != nil {
		f.Close()
		os.Remove(tmp)
		return "", downloadError(spec.RankFileURL, err)
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return "", fmt.Errorf("turbotoken: close temp file: %w", err)
	}

	if err := os.Rename(tmp, localPath); err != nil {
		os.Remove(tmp)
		return "", fmt.Errorf("turbotoken: rename rank file: %w", err)
	}

	return localPath, nil
}

// ReadRankFile reads the rank file bytes for the given encoding,
// downloading the file first if it does not exist locally.
func ReadRankFile(name string) ([]byte, error) {
	path, err := EnsureRankFile(name)
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("turbotoken: read rank file: %w", err)
	}
	return data, nil
}

File: wrappers/rust/src/rank_cache.rs
use std::fs;
use std::path::PathBuf;

use crate::error::TurbotokenError;
use crate::registry::get_encoding_spec;

/// Return the cache directory for rank files.
///
/// Uses `TURBOTOKEN_CACHE_DIR` env if set, otherwise `~/.cache/turbotoken/`.
pub fn cache_dir() -> PathBuf {
    if let Ok(dir) = std::env::var("TURBOTOKEN_CACHE_DIR") {
        return PathBuf::from(dir);
    }
    dirs::cache_dir()
        .unwrap_or_else(|| PathBuf::from("/tmp"))
        .join("turbotoken")
}

/// Ensure a rank file exists locally, downloading it if missing.
/// Returns the path to the cached rank file.
pub fn ensure_rank_file(name: &str) -> Result<PathBuf, TurbotokenError> {
    let spec = get_encoding_spec(name)?;
    let dir = cache_dir();
    let filename = format!("{}.tiktoken", name);
    let path = dir.join(&filename);

    if path.exists() {
        return Ok(path);
    }

    fs::create_dir_all(&dir)?;

    let response = ureq::get(spec.rank_file_url)
        .call()
        .map_err(|e| TurbotokenError::DownloadError(format!("{e}")))?;

    let mut body = Vec::new();
    response
        .into_body()
        .read_to_end(&mut body)
        .map_err(|e| TurbotokenError::DownloadError(format!("read body: {e}")))?;

    // Write atomically via temp file
    let tmp_path = dir.join(format!(".{filename}.tmp"));
    fs::write(&tmp_path, &body)?;
    fs::rename(&tmp_path, &path)?;

    Ok(path)
}

/// Read a rank file's bytes, downloading if needed.
pub fn read_rank_file(name: &str) -> Result<Vec<u8>, TurbotokenError> {
    let path = ensure_rank_file(name)?;
    Ok(fs::read(path)?)
}
